snap/validate_test.go

```diff
@@ -132,6 +154,149 @@ func (s *ValidateSuite) TestValidateAppName(c *C) {
 	}
 }
 
+func (s *ValidateSuite) TestValidateAppSockets(c *C) {
+	app := createSampleApp()
+	app.Sockets["sock"].SocketMode = 0600
+	c.Check(ValidateApp(app), IsNil)
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsEmptyPermsOk(c *C) {
+	app := createSampleApp()
+	c.Check(ValidateApp(app), IsNil)
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsMissingNetworkBindPlug(c *C) {
+	app := createSampleApp()
+	delete(app.Plugs, "network-bind")
+	err := ValidateApp(app)
+	c.Assert(
+		err, ErrorMatches,
+		`"network-bind" interface plug is required when sockets are used`)
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsEmptyListenStream(c *C) {
+	app := createSampleApp()
+	app.Sockets["sock"].ListenStream = ""
+	err := ValidateApp(app)
+	c.Assert(err, ErrorMatches, `socket "sock" must define "listen-stream"`)
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsInvalidName(c *C) {
+	app := createSampleApp()
+	app.Sockets["sock"].Name = "invalid name"
+	err := ValidateApp(app)
+	c.Assert(err, ErrorMatches, `invalid socket name: "invalid name"`)
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsValidListenStreamAddresses(c *C) {
+	app := createSampleApp()
+	validListenAddresses := []string{
+		// socket paths using variables as prefix
+		"$SNAP_DATA/my.socket",
+		"$SNAP_COMMON/my.socket",
+		// abstract sockets
+		"@snap.mysnap.my.socket",
+		// addresses and ports
+		"8080",
+		"127.0.0.1:8080",
+		"[::]:8080",
+		"[::1]:8080",
+	}
+	socket := app.Sockets["sock"]
+	for _, validAddress := range validListenAddresses {
+		socket.ListenStream = validAddress
+		err := ValidateApp(app)
+		c.Assert(err, IsNil)
```

Comment:
in general use `Check` instead of `Assert` to to check things; `Assert` is to check for things that make any other tests pointless (e.g. to check an error value is nil before checking the value returned is as expected).

When doing things in a loop like here, use `Check` and `Commentf`: e.g. `c.Check(err, IsNil, Commentf(validAddress))` to continue testing and have the context of the failing check in the output